### ISSUE TITLE
use of Sub::Attribute cause strange behaviour

### ISSUE THREAD

**Issue by shumphrey** [houseabsolute/test-class-moose#66]
Hello,

I've been tracking down some strange behaviour, and although this isn't Test::Class::Moose's fault, I'm raising this here as Sub::Attribute doesn't seem to be maintained and we observed this issue in Test::Class::Moose first.

The immediate issue is when you introduce a syntax error in a Test::Class::Moose test that error can sometimes not be correctly reported.
This can be demonstrated with two test files.

``` perl
# test.pl
use My::Test;
```

``` perl
# lib/My/Test.pm
use Sub::Attributes;
sub Test : ATTR_SUB {}

sub test_other_routes1 : Test { }
sub test_other_routes2 : Test { }
sub test_other_routes3 : Test { }
sub test_other_routes4 : Test { }
sub test_other_routes5 : Test { }

sub test_that_errors : Test {
    $banana = 1;
}

1;
```

Running <code>perl -c test.pl</code> produces the error message <code>Global symbol "$banana" requires explicit package name (did you forget to declare "my $banana"?)</code> **32 times**. (2^5 ?)
I expect that error message to only be printed once.

Running <code>perl -c lib/My/Test.pm</code> correctly only outputs that error message once.

Replacing the use of Sub::Attribute with native perl does not result in this error:

``` perl

package My::Test;

use strict;
use warnings;

sub MODIFY_CODE_ATTRIBUTES {
    my ($class, $code, @attrs) = @_;

    return ();
}

sub test_other_routes1 : Test { }
sub test_other_routes2 : Test { }
sub test_other_routes3 : Test { }
sub test_other_routes4 : Test { }
sub test_other_routes5 : Test { }

sub test_that_errors : Test {
    $banana = 1;
}

1;
```

Because there seems to be some exponential component to this, with large Test::Class::Moose files we've observed linux grind to a halt with memory consumption and the process get killed.

This behaviour does not occur on older perls, e.g. perl-5.16.3

Looking at the reverse dependencies for Sub::Attribute, it is not used by many things.

**Summary:**
**Test:** <code>perl -c test.pl</code>
**Expected:** Syntax error message to be printed once
**Result:** Syntax error printed 32 times

**Comment by autarch** [comment]
Your code example left out the very critical `use Test::Class::Moose` piece! Please copy examples instead of retyping them.

Once I added that I could reproduce it. That all said, I'm guessing this is either a bug with `Sub::Attributes` or the perl core. I don't know if there's anything TCM can do about it.

**Comment by shumphrey** [comment]
My code was intended not to have <code>use Test::Class::Moose</code> and was a direct copy/paste of my test code.
I believe I stated this was a Sub::Attribute bug not TCM.
However Sub::Attribute appears unmaintained and mostly unused.

**Comment by shumphrey** [comment]
TCM could replace it's use of Sub::Attribute with native Perl to avoid this issue but I can understand not wanting to do that

**Comment by shumphrey** [comment]
It's not a bug in Perl core as stated in ticket.

**Comment by autarch** [comment]
I was only able to trigger the bug by adding `use Test::Class::Moose` to the module code you put in the ticket. Weird.

Anyway, I'd be perfectly happy to get rid of Sub::Attribute if the resulting implementation doesn't end up being too complicated. I'd assume that Attribute::Handlers would work pretty much the same way, or maybe we could just use `attributes` and do this stuff by hand.

@Ovid - did you choose to use Sub::Attribute for a specific reason? The SA docs say that AH won't work if `eval` is used on a sub with attributes, but that seems like quite the corner case.

FWIW, I never use the attribute syntax myself, I prefer to just name everything `sub test_*`. If I wanted to _not_ do that I'd probably just come up with some sort of sugar so I could write something like this:

    test 'something or other' => sub {  ... };

Basically, I avoid attributes like the plague, because all of my past experiences with them have been pretty bad ;)

**Comment by Ovid** [comment]
I can't recall the exact reason I used `Sub::Attribute`, but I know it was because I couldn't get native Perl attributes to work correctly. I _think_ it was because I couldn't `eval` code using attributes because the attributes wouldn't be parsed and thus were ignored.

If that's the case, we have a serious issue if we must have attributes and we must not have `Sub::Attribute`.

**Comment by autarch** [comment]
Coming back to this ...

I have no idea how to resolve these two problems. Personally, I think the "might end up using all our memory because of a buggy unmaintained module" problem is more serious than the "I can't eval code with Test attributes" problem.

**Comment by Ovid** [comment]
@autarch : it's been a long time since I added `Sub::Attribute` to `Test::Class::Moose`. As I recall, I was never able to get native Perl attributes to work, hence `Sub::Attribute`.

**Comment by shumphrey** [comment]
The author of Sub::Attribute is looking for a new maintainer.
https://github.com/gfx/Perl-Sub-Attribute/issues/3

My knowledge of C/XS is insufficient for this.

**Comment by autarch** [comment]
In theory I could possibly maintain it, but I really try to avoid using attributes in code, so I lack motivation ;)

**Comment by DrHyde** [comment]
FYI I've got co-maint on Sub::Attribute. When I get a spare evening I'll release a patched version. First, though, I need to find the reliably failing test I wrote a few weeks back and which seems to have got eaten by a ticketing system somewhere.

**Comment by DrHyde** [comment]
Try https://metacpan.org/release/DCANTRELL/Sub-Attribute-0.05_01. If there's no massive yelling from the CPAN-testers I'll turn it into 0.06 in the next few days.

**Comment by DrHyde** [comment]
Sub::Attribute 0.06 is now on its way to the CPAN.

**Comment by autarch** [comment]
@DrHyde, I can still replicate the "shows a bazillion of the same errors" issue with the latest Sub::Attribute ...

```perl
# This goes in foo.pl
use strict;
use warnings;

use lib 'lib';

use My::Test;

1;
```

and then ...

```perl
# This goes in lib/My/Test.pm
package My::Test;

use strict;
use warnings;

use Test::Class::Moose;

sub test_other_routes1 : Test { }
sub test_other_routes2 : Test { }
sub test_other_routes3 : Test { }
sub test_other_routes4 : Test { }
sub test_other_routes5 : Test { }

sub test_that_errors : Test {
    $banana = 1;
}

1;
```

**Comment by autarch** [comment]
Also not specific to TCM:

```perl
package My::Test;

use strict;
use warnings;

#use Test::Class::Moose;
use Sub::Attribute;

sub Test :ATTR_SUB{}

sub test_other_routes1 : Test { }
sub test_other_routes2 : Test { }
sub test_other_routes3 : Test { }
sub test_other_routes4 : Test { }
sub test_other_routes5 : Test { }

sub test_that_errors : Test {
    $banana = 1;
}

1;
```

**Comment by autarch** [comment]
This appears to be fixed in Sub::Attribute 0.06. Thanks @DrHyde!
